pep517/build.py

```diff
@@ -3,25 +3,50 @@
 import argparse
 import logging
 import os
-import contextlib
 import pytoml
 import shutil
-import errno
-import tempfile
 
 from .envbuild import BuildEnvironment
 from .wrappers import Pep517HookCaller
+from .dirtools import tempdir, mkdir_p
 
 log = logging.getLogger(__name__)
 
 
-@contextlib.contextmanager
-def tempdir():
-    td = tempfile.mkdtemp()
+def validate_system(system):
+    """
+    Ensure build system has the requisite fields.
+    """
+    required = {'requires', 'build-backend'}
+    if required > set(system):
```

Comment:
I don't think this does quite what you expect. Try

```python
if not (required <= set(system)):
```

`>` catches if the keys are a strict subset of the required ones. So it works if you miss one out completely, but if you miss one out and add another - e.g. by making a typo - the keys are no longer a subset.

`<=` checks that the keys are a non-strict superset of the required ones, which is what we care about.

Reply:
Good catch. I was able to replicate your findings with this patch:

```
diff --git a/conftest.py b/conftest.py
new file mode 100644
index 0000000..4f1eca8
--- /dev/null
+++ b/conftest.py
@@ -0,0 +1,7 @@
+import sys
+
+collect_ignore = ['pep517/_in_process.py']
+
+
+if sys.version_info < (3,):
+    collect_ignore.append('pep517/build.py')
diff --git a/pep517/build.py b/pep517/build.py
index 781f450..8aafc1c 100644
--- a/pep517/build.py
+++ b/pep517/build.py
@@ -17,6 +17,31 @@ log = logging.getLogger(__name__)
 def validate_system(system):
     """
     Ensure build system has the requisite fields.
+
+    >>> validate_system({'build-backend': None, 'requires': None})
+    >>> validate_system({'build-backend': None, 'requires': None,
+    ...     'other': None})
+
+    >>> validate_system(dict(requires=None))
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'build-backend'}
+
+    >>> validate_system({'build-backend': None})
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'requires'}
+
+    >>> validate_system({})
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'...', '...'}
+
+    >>> validate_system({'build-backend': None, 'other': None})
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'requires'}
+
     """
     required = {'requires', 'build-backend'}
     if required > set(system):
diff --git a/pytest.ini b/pytest.ini
index 5c37966..b5835b8 100644
--- a/pytest.ini
+++ b/pytest.ini
@@ -1,2 +1,2 @@
 [pytest]
-addopts=--flake8
+addopts=--flake8 --doctest-modules
```

(only runs on Python 3)

However, there are too many complications with using doctests on this project, so I'm abandoning that work.

Reply:
Fixed in 1f90e57.

Reply:
@jaraco 

Just passing through, following a link from "GitHub Explore the week", and having decided to concentrate more on doctests (at work): may I ask what complications arose here with using them?